wallet/app-vty/Glue.hs

```diff
@@ -0,0 +1,287 @@
+              writeIORef walletStateRef $ walletState{ selection = mfilter ((< length (accounts walletState)) . fromIntegral) $ Just i }
+              walletRefreshState
+              return $ Left "No further action"
+          | otherwise -> do
+              writeIORef walletStateRef $ walletState{ selection = Nothing }
+              walletRefreshState
+              return $ Left "No further action"
+        _ -> case opToExpr walletState op of
+          Left err -> return $ Left err
+          Right expr -> fmap Right $ putCommand (uiCommandHandle op) expr
+    uiCommandHandle op commandId = KnitCommandHandle
+      { putCommandResult = \mtid result ->
+          whenJust (resultToUI result op) $ putUiEvent . UiCommandResult (commandIdToUI commandId mtid)
+      , putCommandOutput = \_ _ ->
+          return ()
+      }
+
+    optString key value = if null value then [] else [Knit.ArgKw key . Knit.ExprLit . Knit.toLit . Knit.LitString $ value]
+
+    opToExpr WalletState{..} = \case
+      UiBalance -> do
+        account <- maybeToRight "No account selected" $ selection >>= (accounts ^?) . ix . fromIntegral
+        Right $ Knit.ExprProcCall
+          (Knit.ProcCall "get-balance"
+            [Knit.ArgPos . Knit.ExprLit . Knit.toLit . Knit.LitAddress . accountAddress $ account]
+          )
+      UiSend UiSendArgs{..} -> do
+        account <- maybeToRight "No account selected" $ selection >>= (accounts ^?) . ix . fromIntegral
+        argOutputs <- forM usaOutputs $ \UiSendOutput{..} -> do
+          argAddress <- addrFromText usoAddress
+          argCoin <- readEither usoAmount
+          Right $ Knit.ArgKw "out" . Knit.ExprProcCall $ Knit.ProcCall "tx-out"
+            [ Knit.ArgPos . Knit.ExprLit . Knit.toLit . Knit.LitAddress $ argAddress
+            , Knit.ArgPos . Knit.ExprLit . Knit.toLit . Knit.LitNumber $ argCoin
+            ]
+        Right $ Knit.ExprProcCall
+          (Knit.ProcCall "send-tx" $
+            [ Knit.ArgPos . Knit.ExprLit . Knit.toLit . Knit.LitString . show . accountSecretKey $ account ] ++
+            argOutputs ++
+            optString "pass" usaPassphrase
+          )
+      UiKill commandId ->
+        Right $ Knit.ExprProcCall
+          (Knit.ProcCall Knit.killCommandName
+            [Knit.ArgPos . Knit.ExprLit . Knit.toLit . Knit.LitTaskId . TaskId $ commandId]
+          )
+      UiNewWallet UiNewWalletArgs{..} -> do
+        Right $ Knit.ExprProcCall
+          (Knit.ProcCall "gen-key-pair" $
+            optString "name" unwaName ++
+            optString "pass" unwaPassphrase
+          )
+      UiRestoreWallet UiRestoreWalletArgs{..} -> do
+        Right $ Knit.ExprProcCall
+          (Knit.ProcCall "restore-key" $
+            [ Knit.ArgKw "secretkey" . Knit.ExprLit . Knit.toLit . Knit.LitString $ urwaMnemonic
+            ] ++
+            optString "name" urwaName ++
+            optString "pass" urwaPassphrase
+          )
+      _ -> Left "Not implemented yet"
```

Comment:
By the way, which commands are missing now?

Reply:
New account, new address (for HD-wallets), rename, remove...

Reply:
Ah ok, these constructors are imported (copypasted) from ariadne i guess?

Reply:
Yes.